sno/diff_output.py

```diff
@@ -495,29 +524,47 @@ def _out(dataset, ds_diff):
 class LazyJsonFeatureOutput:
     """Wrapper of KeyValue that lazily serialises it as JSON when sent to json.dumps"""
 
-    __slots__ = "key_value"
+    __slots__ = ("key_value", "geometry_transform")
 
-    def __init__(self, key_value):
+    def __init__(self, key_value, geometry_transform=None):
         self.key_value = key_value
+        self.geometry_transform = geometry_transform
 
     def __json__(self):
-        return json_row(self.key_value.get_lazy_value())
+        return json_row(
+            self.key_value.get_lazy_value(), geometry_transform=self.geometry_transform
+        )
 
 
 @ungenerator(dict)
-def json_row(row):
+def json_row(row, geometry_transform=None):
     """
     Turns a row into a dict for serialization as JSON.
     The geometry is serialized as hexWKB.
     """
     for k, v in row.items():
         if isinstance(v, bytes):
-            v = gpkg_geom_to_hex_wkb(v)
+            if geometry_transform is None:
+                v = gpkg_geom_to_hex_wkb(v)
+            else:
+                # reproject
+                ogr_geom = gpkg_geom_to_ogr(v)
+                ogr_geom.Transform(geometry_transform)
```

Comment:
does this raise exceptions correctly?

Reply:
```points2 (BARE:master) $ sno diff -o json | grep geom
            "geom": "POINT (173.831599213632 -87.1267244418588)",
points2 (BARE:master) $ sno diff -o json --crs epsg:4326 | grep geom
            "geom": "POINT (173.831599213632 -87.1267244418588)",
points2 (BARE:master) $ sno diff -o json --crs epsg:3857 | grep geom
            "geom": "POINT (19350845.108242 -23507975.1660679)",
```

Reply:
There doesn't seem to be a way to make this fail. It does eventually produce `inf` values, but even detecting that will be slow on our end and thus not worth it. I can't see a way to make OGR/proj reject these.

Reply:
From https://github.com/OSGeo/PROJ/blob/master/test/gie/builtins.gie#L2548-L2549

```
>>> from osgeo import osr, ogr
>>> osr.UseExceptions()
>>> ogr.UseExceptions()
>>> srs = osr.SpatialReference()
>>> srs.SetFromUserInput("+proj=laea +ellps=GRS80")
0
>>> g_ok = ogr.CreateGeometryFromWkt("POINT(2 1)")
>>> g_err = ogr.CreateGeometryFromWkt("POINT(180 0)")
>>> ll = osr.SpatialReference()
>>> ll.SetWellKnownGeogCS('WGS84')
0
>>> tx = osr.CreateCoordinateTransformation(ll, srs)
>>> g_ok.Transform(tx)
0
>>> g_ok.ExportToWkt()
'POINT (111267.664536143 221146.644064019)'
>>> g_err.Transform(tx)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/osgeo/ogr.py", line 7791, in Transform
    return _ogr.Geometry_Transform(self, *args)
RuntimeError: OGR Error: General Error
```

Reply:
OK. That happens here too. I'll wrap it with InvalidOperation. One awkward thing is that errors here happen halfway through a JSON stream. But there's not much we can do about that apart from processing everything up front, which we're trying hard to avoid.

Reply:
Sure, I think that's somewhat inevitable.

Though, modern proj knows a _lot_ more about coordinate systems, areas of use, etc in order to do the best possible transform...

https://proj.org/faq.html#can-i-transform-from-abc-to-xyz
https://proj.org/operations/operations_computation.html

Wonder if we can do the equivalent of that in GDAL and issue a warning if it's using Ballpark transforms?

```console
$ projinfo --summary -s 'epsg:27200' -t 'epsg:3416'  # 3416 = ETRS89 / Austria Lambert
Candidate operations found: 1
unknown id, Inverse of New Zealand Map Grid + Ballpark geographic offset from NZGD49 to ETRS89 + Austria Lambert, unknown accuracy, World, has ballpark transformation
$ projinfo --summary -s 'epsg:27200' -t 'epsg:3416' --hide-ballpark
Candidate operations found: 0
```